File: elixir_ring/lib/ring_system/perf_tracker.ex
defmodule RingSystem.PerfTracker do
  @moduledoc """
  Collects per-token latencies and computes summary statistics.
  """

  use Agent

  def start_link do
    Agent.start_link(fn ->
      %{latencies: [], start_time: System.monotonic_time(:microsecond)}
    end)
  end

  def record_latency(pid, latency_us) do
    Agent.update(pid, fn state ->
      %{state | latencies: [latency_us | state.latencies]}
    end)
  end

  def print_summary(pid) do
    state = Agent.get(pid, & &1)
    latencies = Enum.sort(state.latencies)
    count = length(latencies)

    if count == 0 do
      IO.puts("\n[PerfTracker] No tokens processed.")
    else
      elapsed_us = System.monotonic_time(:microsecond) - state.start_time
      elapsed_s = elapsed_us / 1_000_000

      min_lat = hd(latencies)
      max_lat = List.last(latencies)
      avg_lat = Enum.sum(latencies) / count
      p50 = percentile(latencies, count, 50)
      p95 = percentile(latencies, count, 95)
      throughput = count / elapsed_s

      IO.puts("\n[PerfTracker] === Performance Summary ===")
      IO.puts("[PerfTracker] Tokens processed: #{count}")
      IO.puts("[PerfTracker] Elapsed time: #{Float.round(elapsed_s, 3)}s")
      IO.puts("[PerfTracker] Throughput: #{Float.round(throughput, 1)} tokens/sec")
      IO.puts("[PerfTracker] Latency (us): min=#{min_lat}, max=#{max_lat}, avg=#{round(avg_lat)}")
      IO.puts("[PerfTracker] Latency (us): p50=#{p50}, p95=#{p95}")
    end
  end

  defp percentile(sorted, count, p) do
    idx = max(0, round(count * p / 100) - 1)
    Enum.at(sorted, idx)
  end
end

File: java_ring/src/main/java/ringsystem/PerfTracker.java
package ringsystem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PerfTracker {
    private final List<Long> latencies = Collections.synchronizedList(new ArrayList<>());
    private final long startTimeNanos = System.nanoTime();

    public void recordLatency(long latencyUs) {
        latencies.add(latencyUs);
    }

    public void printSummary() {
        List<Long> sorted;
        synchronized (latencies) {
            sorted = new ArrayList<>(latencies);
        }
        Collections.sort(sorted);
        int count = sorted.size();

        if (count == 0) {
            System.out.println("\n[PerfTracker] No tokens processed.");
            return;
        }

        long elapsedUs = (System.nanoTime() - startTimeNanos) / 1000;
        double elapsedS = elapsedUs / 1_000_000.0;

        long min = sorted.get(0);
        long max = sorted.get(count - 1);
        long sum = sorted.stream().mapToLong(Long::longValue).sum();
        long avg = sum / count;
        long p50 = sorted.get(Math.max(0, (int) (count * 0.50) - 1));
        long p95 = sorted.get(Math.max(0, (int) (count * 0.95) - 1));
        double throughput = count / elapsedS;

        System.out.println("\n[PerfTracker] === Performance Summary ===");
        System.out.printf("[PerfTracker] Tokens processed: %d%n", count);
        System.out.printf("[PerfTracker] Elapsed time: %.3fs%n", elapsedS);
        System.out.printf("[PerfTracker] Throughput: %.1f tokens/sec%n", throughput);
        System.out.printf("[PerfTracker] Latency (us): min=%d, max=%d, avg=%d%n", min, max, avg);
        System.out.printf("[PerfTracker] Latency (us): p50=%d, p95=%d%n", p50, p95);
    }
}
